// @tinymce/tinymce-angular/bundles/tinymce-tinymce-angular.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('@tinymce/tinymce-angular', ['exports', '@angular/core', '@angular/common', '@angular/forms'], factory) :
    (global = global || self, factory((global.tinymce = global.tinymce || {}, global.tinymce['tinymce-angular'] = {}), global.ng.core, global.ng.common, global.ng.forms));
}(this, (function (exports, core, common, forms) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * Copyright (c) 2017-present, Ephox, Inc.
     *
     * This source code is licensed under the Apache 2 license found in the
     * LICENSE file in the root directory of this source tree.
     *
     */
    var getTinymce = function () {
        var w = typeof window !== 'undefined' ? window : undefined;
        return w && w.tinymce ? w.tinymce : null;
    };
    var ɵ0 = getTinymce;

    var Events = /** @class */ (function () {
        function Events() {
            this.onBeforePaste = new core.EventEmitter();
            this.onBlur = new core.EventEmitter();
            this.onClick = new core.EventEmitter();
            this.onContextMenu = new core.EventEmitter();
            this.onCopy = new core.EventEmitter();
            this.onCut = new core.EventEmitter();
            this.onDblclick = new core.EventEmitter();
            this.onDrag = new core.EventEmitter();
            this.onDragDrop = new core.EventEmitter();
            this.onDragEnd = new core.EventEmitter();
            this.onDragGesture = new core.EventEmitter();
            this.onDragOver = new core.EventEmitter();
            this.onDrop = new core.EventEmitter();
            this.onFocus = new core.EventEmitter();
            this.onFocusIn = new core.EventEmitter();
            this.onFocusOut = new core.EventEmitter();
            this.onKeyDown = new core.EventEmitter();
            this.onKeyPress = new core.EventEmitter();
            this.onKeyUp = new core.EventEmitter();
            this.onMouseDown = new core.EventEmitter();
            this.onMouseEnter = new core.EventEmitter();
            this.onMouseLeave = new core.EventEmitter();
            this.onMouseMove = new core.EventEmitter();
            this.onMouseOut = new core.EventEmitter();
            this.onMouseOver = new core.EventEmitter();
            this.onMouseUp = new core.EventEmitter();
            this.onPaste = new core.EventEmitter();
            this.onSelectionChange = new core.EventEmitter();
            this.onActivate = new core.EventEmitter();
            this.onAddUndo = new core.EventEmitter();
            this.onBeforeAddUndo = new core.EventEmitter();
            this.onBeforeExecCommand = new core.EventEmitter();
            this.onBeforeGetContent = new core.EventEmitter();
            this.onBeforeRenderUI = new core.EventEmitter();
            this.onBeforeSetContent = new core.EventEmitter();
            this.onChange = new core.EventEmitter();
            this.onClearUndos = new core.EventEmitter();
            this.onDeactivate = new core.EventEmitter();
            this.onDirty = new core.EventEmitter();
            this.onExecCommand = new core.EventEmitter();
            this.onGetContent = new core.EventEmitter();
            this.onHide = new core.EventEmitter();
            this.onInit = new core.EventEmitter();
            this.onLoadContent = new core.EventEmitter();
            this.onNodeChange = new core.EventEmitter();
            this.onPostProcess = new core.EventEmitter();
            this.onPostRender = new core.EventEmitter();
            this.onPreInit = new core.EventEmitter();
            this.onPreProcess = new core.EventEmitter();
            this.onProgressState = new core.EventEmitter();
            this.onRedo = new core.EventEmitter();
            this.onRemove = new core.EventEmitter();
            this.onReset = new core.EventEmitter();
            this.onSaveContent = new core.EventEmitter();
            this.onSetAttrib = new core.EventEmitter();
            this.onObjectResizeStart = new core.EventEmitter();
            this.onObjectResized = new core.EventEmitter();
            this.onObjectSelected = new core.EventEmitter();
            this.onSetContent = new core.EventEmitter();
            this.onShow = new core.EventEmitter();
            this.onSubmit = new core.EventEmitter();
            this.onUndo = new core.EventEmitter();
            this.onVisualAid = new core.EventEmitter();
        }
        __decorate([
            core.Output()
        ], Events.prototype, "onBeforePaste", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onBlur", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onClick", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onContextMenu", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onCopy", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onCut", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDblclick", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDrag", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDragDrop", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDragEnd", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDragGesture", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDragOver", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDrop", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onFocus", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onFocusIn", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onFocusOut", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onKeyDown", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onKeyPress", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onKeyUp", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onMouseDown", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onMouseEnter", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onMouseLeave", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onMouseMove", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onMouseOut", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onMouseOver", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onMouseUp", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onPaste", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onSelectionChange", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onActivate", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onAddUndo", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onBeforeAddUndo", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onBeforeExecCommand", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onBeforeGetContent", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onBeforeRenderUI", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onBeforeSetContent", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onChange", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onClearUndos", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDeactivate", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onDirty", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onExecCommand", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onGetContent", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onHide", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onInit", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onLoadContent", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onNodeChange", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onPostProcess", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onPostRender", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onPreInit", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onPreProcess", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onProgressState", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onRedo", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onRemove", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onReset", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onSaveContent", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onSetAttrib", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onObjectResizeStart", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onObjectResized", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onObjectSelected", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onSetContent", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onShow", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onSubmit", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onUndo", void 0);
        __decorate([
            core.Output()
        ], Events.prototype, "onVisualAid", void 0);
        return Events;
    }());
    var validEvents = [
        'onActivate',
        'onAddUndo',
        'onBeforeAddUndo',
        'onBeforeExecCommand',
        'onBeforeGetContent',
        'onBeforeRenderUI',
        'onBeforeSetContent',
        'onBeforePaste',
        'onBlur',
        'onChange',
        'onClearUndos',
        'onClick',
        'onContextMenu',
        'onCopy',
        'onCut',
        'onDblclick',
        'onDeactivate',
        'onDirty',
        'onDrag',
        'onDragDrop',
        'onDragEnd',
        'onDragGesture',
        'onDragOver',
        'onDrop',
        'onExecCommand',
        'onFocus',
        'onFocusIn',
        'onFocusOut',
        'onGetContent',
        'onHide',
        'onInit',
        'onKeyDown',
        'onKeyPress',
        'onKeyUp',
        'onLoadContent',
        'onMouseDown',
        'onMouseEnter',
        'onMouseLeave',
        'onMouseMove',
        'onMouseOut',
        'onMouseOver',
        'onMouseUp',
        'onNodeChange',
        'onObjectResizeStart',
        'onObjectResized',
        'onObjectSelected',
        'onPaste',
        'onPostProcess',
        'onPostRender',
        'onPreProcess',
        'onProgressState',
        'onRedo',
        'onRemove',
        'onReset',
        'onSaveContent',
        'onSelectionChange',
        'onSetAttrib',
        'onSetContent',
        'onShow',
        'onSubmit',
        'onUndo',
        'onVisualAid'
    ];

    /**
     * Copyright (c) 2017-present, Ephox, Inc.
     *
     * This source code is licensed under the Apache 2 license found in the
     * LICENSE file in the root directory of this source tree.
     *
     */
    var bindHandlers = function (ctx, editor) {
        validEvents.forEach(function (eventName) {
            var eventEmitter = ctx[eventName];
            editor.on(eventName.substring(2), function (event) { return ctx.ngZone.run(function () { return eventEmitter.emit({ event: event, editor: editor }); }); });
        });
    };
    var ɵ0$1 = bindHandlers;
    var unique = 0;
    var uuid = function (prefix) {
        var date = new Date();
        var time = date.getTime();
        var random = Math.floor(Math.random() * 1000000000);
        unique++;
        return prefix + '_' + random + unique + String(time);
    };
    var ɵ1 = uuid;
    var isTextarea = function (element) {
        return typeof element !== 'undefined' && element.tagName.toLowerCase() === 'textarea';
    };
    var ɵ2 = isTextarea;
    var normalizePluginArray = function (plugins) {
        if (typeof plugins === 'undefined' || plugins === '') {
            return [];
        }
        return Array.isArray(plugins) ? plugins : plugins.split(' ');
    };
    var ɵ3 = normalizePluginArray;
    var mergePlugins = function (initPlugins, inputPlugins) {
        return normalizePluginArray(initPlugins).concat(normalizePluginArray(inputPlugins));
    };
    var ɵ4 = mergePlugins;
    // tslint:disable-next-line:no-empty
    var noop = function () { };
    var ɵ5 = noop;
    var isNullOrUndefined = function (value) { return value === null || value === undefined; };
    var ɵ6 = isNullOrUndefined;

    /**
     * Copyright (c) 2017-present, Ephox, Inc.
     *
     * This source code is licensed under the Apache 2 license found in the
     * LICENSE file in the root directory of this source tree.
     *
     */
    var createState = function () {
        return {
            listeners: [],
            scriptId: uuid('tiny-script'),
            scriptLoaded: false
        };
    };
    var ɵ0$2 = createState;
    var CreateScriptLoader = function () {
        var state = createState();
        var injectScriptTag = function (scriptId, doc, url, callback) {
            var scriptTag = doc.createElement('script');
            scriptTag.referrerPolicy = 'origin';
            scriptTag.type = 'application/javascript';
            scriptTag.id = scriptId;
            scriptTag.src = url;
            var handler = function () {
                scriptTag.removeEventListener('load', handler);
                callback();
            };
            scriptTag.addEventListener('load', handler);
            if (doc.head) {
                doc.head.appendChild(scriptTag);
            }
        };
        var load = function (doc, url, callback) {
            if (state.scriptLoaded) {
                callback();
            }
            else {
                state.listeners.push(callback);
                if (!doc.getElementById(state.scriptId)) {
                    injectScriptTag(state.scriptId, doc, url, function () {
                        state.listeners.forEach(function (fn) { return fn(); });
                        state.scriptLoaded = true;
                    });
                }
            }
        };
        // Only to be used by tests.
        var reinitialize = function () {
            state = createState();
        };
        return {
            load: load,
            reinitialize: reinitialize
        };
    };
    var ɵ1$1 = CreateScriptLoader;
    var ScriptLoader = CreateScriptLoader();

    var TINYMCE_SCRIPT_SRC = new core.InjectionToken('TINYMCE_SCRIPT_SRC');
    var EDITOR_COMPONENT_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        useExisting: core.forwardRef(function () { return EditorComponent; }),
        multi: true
    };
    var EditorComponent = /** @class */ (function (_super) {
        __extends(EditorComponent, _super);
        function EditorComponent(elementRef, ngZone, platformId, tinymceScriptSrc) {
            var _this = _super.call(this) || this;
            _this.platformId = platformId;
            _this.tinymceScriptSrc = tinymceScriptSrc;
            _this.cloudChannel = '5';
            _this.apiKey = 'no-api-key';
            _this.id = '';
            _this.modelEvents = 'change keyup undo redo';
            _this.onTouchedCallback = noop;
            _this.onChangeCallback = noop;
            _this._elementRef = elementRef;
            _this.ngZone = ngZone;
            _this.initialise = _this.initialise.bind(_this);
            return _this;
        }
        Object.defineProperty(EditorComponent.prototype, "disabled", {
            get: function () {
                return this._disabled;
            },
            set: function (val) {
                this._disabled = val;
                if (this._editor && this._editor.initialized) {
                    this._editor.setMode(val ? 'readonly' : 'design');
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(EditorComponent.prototype, "editor", {
            get: function () {
                return this._editor;
            },
            enumerable: true,
            configurable: true
        });
        EditorComponent.prototype.writeValue = function (value) {
            if (this._editor && this._editor.initialized) {
                this._editor.setContent(isNullOrUndefined(value) ? '' : value);
            }
            else {
                this.initialValue = value === null ? undefined : value;
            }
        };
        EditorComponent.prototype.registerOnChange = function (fn) {
            this.onChangeCallback = fn;
        };
        EditorComponent.prototype.registerOnTouched = function (fn) {
            this.onTouchedCallback = fn;
        };
        EditorComponent.prototype.setDisabledState = function (isDisabled) {
            if (this._editor) {
                this._editor.setMode(isDisabled ? 'readonly' : 'design');
            }
            else if (isDisabled) {
                this.init = __assign({}, this.init, { readonly: true });
            }
        };
        EditorComponent.prototype.ngAfterViewInit = function () {
            if (common.isPlatformBrowser(this.platformId)) {
                this.id = this.id || uuid('tiny-angular');
                this.inline =
                    typeof this.inline !== 'undefined' ? (typeof this.inline === 'boolean' ? this.inline : true) : this.init && this.init.inline;
                this.createElement();
                if (getTinymce() !== null) {
                    this.initialise();
                }
                else if (this._element && this._element.ownerDocument) {
                    ScriptLoader.load(this._element.ownerDocument, this.getScriptSrc(), this.initialise);
                }
            }
        };
        EditorComponent.prototype.ngOnDestroy = function () {
            if (getTinymce() !== null) {
                getTinymce().remove(this._editor);
            }
        };
        EditorComponent.prototype.createElement = function () {
            var tagName = typeof this.tagName === 'string' ? this.tagName : 'div';
            this._element = document.createElement(this.inline ? tagName : 'textarea');
            if (this._element) {
                this._element.id = this.id;
                if (isTextarea(this._element)) {
                    this._element.style.visibility = 'hidden';
                }
                this._elementRef.nativeElement.appendChild(this._element);
            }
        };
        EditorComponent.prototype.initialise = function () {
            var _this = this;
            var finalInit = __assign({}, this.init, { target: this._element, inline: this.inline, readonly: this.disabled, plugins: mergePlugins(this.init && this.init.plugins, this.plugins), toolbar: this.toolbar || (this.init && this.init.toolbar), setup: function (editor) {
                    _this._editor = editor;
                    editor.on('init', function (e) {
                        _this.initEditor(editor);
                    });
                    bindHandlers(_this, editor);
                    if (_this.init && typeof _this.init.setup === 'function') {
                        _this.init.setup(editor);
                    }
                } });
            if (isTextarea(this._element)) {
                this._element.style.visibility = '';
            }
            this.ngZone.runOutsideAngular(function () {
                getTinymce().init(finalInit);
            });
        };
        EditorComponent.prototype.getScriptSrc = function () {
            return isNullOrUndefined(this.tinymceScriptSrc) ?
                "https://cdn.tiny.cloud/1/" + this.apiKey + "/tinymce/" + this.cloudChannel + "/tinymce.min.js" :
                this.tinymceScriptSrc;
        };
        EditorComponent.prototype.initEditor = function (editor) {
            var _this = this;
            editor.on('blur', function () { return _this.ngZone.run(function () { return _this.onTouchedCallback(); }); });
            editor.on(this.modelEvents, function () {
                _this.ngZone.run(function () { return _this.onChangeCallback(editor.getContent({ format: _this.outputFormat })); });
            });
            if (typeof this.initialValue === 'string') {
                this.ngZone.run(function () {
                    editor.setContent(_this.initialValue);
                    _this.onChangeCallback(editor.getContent({ format: _this.outputFormat }));
                });
            }
        };
        EditorComponent.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.NgZone },
            { type: Object, decorators: [{ type: core.Inject, args: [core.PLATFORM_ID,] }] },
            { type: String, decorators: [{ type: core.Optional }, { type: core.Inject, args: [TINYMCE_SCRIPT_SRC,] }] }
        ]; };
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "disabled", null);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "cloudChannel", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "apiKey", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "init", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "id", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "initialValue", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "outputFormat", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "inline", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "tagName", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "plugins", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "toolbar", void 0);
        __decorate([
            core.Input()
        ], EditorComponent.prototype, "modelEvents", void 0);
        EditorComponent = __decorate([
            core.Component({
                selector: 'editor',
                template: '<ng-template></ng-template>',
                providers: [EDITOR_COMPONENT_VALUE_ACCESSOR],
                styles: [':host { display: block; }']
            }),
            __param(2, core.Inject(core.PLATFORM_ID)),
            __param(3, core.Optional()), __param(3, core.Inject(TINYMCE_SCRIPT_SRC))
        ], EditorComponent);
        return EditorComponent;
    }(Events));

    var EditorModule = /** @class */ (function () {
        function EditorModule() {
        }
        EditorModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, forms.FormsModule],
                declarations: [EditorComponent],
                exports: [EditorComponent]
            })
        ], EditorModule);
        return EditorModule;
    }());

    exports.EditorComponent = EditorComponent;
    exports.EditorModule = EditorModule;
    exports.TINYMCE_SCRIPT_SRC = TINYMCE_SCRIPT_SRC;
    exports.ɵa = Events;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
