// @tinymce/tinymce-angular/esm5/editor/Events.js
import * as tslib_1 from "tslib";
import { Output, EventEmitter } from '@angular/core';
var Events = /** @class */ (function () {
    function Events() {
        this.onBeforePaste = new EventEmitter();
        this.onBlur = new EventEmitter();
        this.onClick = new EventEmitter();
        this.onContextMenu = new EventEmitter();
        this.onCopy = new EventEmitter();
        this.onCut = new EventEmitter();
        this.onDblclick = new EventEmitter();
        this.onDrag = new EventEmitter();
        this.onDragDrop = new EventEmitter();
        this.onDragEnd = new EventEmitter();
        this.onDragGesture = new EventEmitter();
        this.onDragOver = new EventEmitter();
        this.onDrop = new EventEmitter();
        this.onFocus = new EventEmitter();
        this.onFocusIn = new EventEmitter();
        this.onFocusOut = new EventEmitter();
        this.onKeyDown = new EventEmitter();
        this.onKeyPress = new EventEmitter();
        this.onKeyUp = new EventEmitter();
        this.onMouseDown = new EventEmitter();
        this.onMouseEnter = new EventEmitter();
        this.onMouseLeave = new EventEmitter();
        this.onMouseMove = new EventEmitter();
        this.onMouseOut = new EventEmitter();
        this.onMouseOver = new EventEmitter();
        this.onMouseUp = new EventEmitter();
        this.onPaste = new EventEmitter();
        this.onSelectionChange = new EventEmitter();
        this.onActivate = new EventEmitter();
        this.onAddUndo = new EventEmitter();
        this.onBeforeAddUndo = new EventEmitter();
        this.onBeforeExecCommand = new EventEmitter();
        this.onBeforeGetContent = new EventEmitter();
        this.onBeforeRenderUI = new EventEmitter();
        this.onBeforeSetContent = new EventEmitter();
        this.onChange = new EventEmitter();
        this.onClearUndos = new EventEmitter();
        this.onDeactivate = new EventEmitter();
        this.onDirty = new EventEmitter();
        this.onExecCommand = new EventEmitter();
        this.onGetContent = new EventEmitter();
        this.onHide = new EventEmitter();
        this.onInit = new EventEmitter();
        this.onLoadContent = new EventEmitter();
        this.onNodeChange = new EventEmitter();
        this.onPostProcess = new EventEmitter();
        this.onPostRender = new EventEmitter();
        this.onPreInit = new EventEmitter();
        this.onPreProcess = new EventEmitter();
        this.onProgressState = new EventEmitter();
        this.onRedo = new EventEmitter();
        this.onRemove = new EventEmitter();
        this.onReset = new EventEmitter();
        this.onSaveContent = new EventEmitter();
        this.onSetAttrib = new EventEmitter();
        this.onObjectResizeStart = new EventEmitter();
        this.onObjectResized = new EventEmitter();
        this.onObjectSelected = new EventEmitter();
        this.onSetContent = new EventEmitter();
        this.onShow = new EventEmitter();
        this.onSubmit = new EventEmitter();
        this.onUndo = new EventEmitter();
        this.onVisualAid = new EventEmitter();
    }
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onBeforePaste", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onBlur", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onClick", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onContextMenu", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onCopy", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onCut", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDblclick", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDrag", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDragDrop", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDragEnd", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDragGesture", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDragOver", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDrop", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onFocus", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onFocusIn", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onFocusOut", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onKeyDown", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onKeyPress", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onKeyUp", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onMouseDown", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onMouseEnter", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onMouseLeave", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onMouseMove", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onMouseOut", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onMouseOver", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onMouseUp", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onPaste", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onSelectionChange", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onActivate", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onAddUndo", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onBeforeAddUndo", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onBeforeExecCommand", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onBeforeGetContent", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onBeforeRenderUI", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onBeforeSetContent", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onChange", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onClearUndos", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDeactivate", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onDirty", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onExecCommand", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onGetContent", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onHide", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onInit", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onLoadContent", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onNodeChange", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onPostProcess", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onPostRender", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onPreInit", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onPreProcess", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onProgressState", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onRedo", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onRemove", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onReset", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onSaveContent", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onSetAttrib", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onObjectResizeStart", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onObjectResized", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onObjectSelected", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onSetContent", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onShow", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onSubmit", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onUndo", void 0);
    tslib_1.__decorate([
        Output()
    ], Events.prototype, "onVisualAid", void 0);
    return Events;
}());
export { Events };
export var validEvents = [
    'onActivate',
    'onAddUndo',
    'onBeforeAddUndo',
    'onBeforeExecCommand',
    'onBeforeGetContent',
    'onBeforeRenderUI',
    'onBeforeSetContent',
    'onBeforePaste',
    'onBlur',
    'onChange',
    'onClearUndos',
    'onClick',
    'onContextMenu',
    'onCopy',
    'onCut',
    'onDblclick',
    'onDeactivate',
    'onDirty',
    'onDrag',
    'onDragDrop',
    'onDragEnd',
    'onDragGesture',
    'onDragOver',
    'onDrop',
    'onExecCommand',
    'onFocus',
    'onFocusIn',
    'onFocusOut',
    'onGetContent',
    'onHide',
    'onInit',
    'onKeyDown',
    'onKeyPress',
    'onKeyUp',
    'onLoadContent',
    'onMouseDown',
    'onMouseEnter',
    'onMouseLeave',
    'onMouseMove',
    'onMouseOut',
    'onMouseOver',
    'onMouseUp',
    'onNodeChange',
    'onObjectResizeStart',
    'onObjectResized',
    'onObjectSelected',
    'onPaste',
    'onPostProcess',
    'onPostRender',
    'onPreProcess',
    'onProgressState',
    'onRedo',
    'onRemove',
    'onReset',
    'onSaveContent',
    'onSelectionChange',
    'onSetAttrib',
    'onSetContent',
    'onShow',
    'onSubmit',
    'onUndo',
    'onVisualAid'
];
